ocaml/xapi/xapi_xenops.ml

```diff
@@ -2058,7 +2058,15 @@ let update_task ~__context queue_name id =
     let task_t = Client.TASK.stat dbg id in
     match task_t.Task.state with
     | Task.Pending x ->
-      Db.Task.set_progress ~__context ~self ~value:x
+      Db.Task.set_progress ~__context ~self ~value:x;
+      if not task_t.Task.cancellable then begin
+        let allowed_operations = Db.Task.get_allowed_operations ~__context ~self in
+        if List.mem `cancel allowed_operations then begin
+          let allowed_operations' = List.filter (fun x -> x != `cancel) allowed_operations in
+          debug "Set task %s to not cancellable." (Ref.string_of self);
```

Comment:
Here you can use `Ref.really_pretty_and_small` to get a task ref as in the logs (e.g. `R:xxxxxxxx`), which makes it easier to follow.

Reply:
Does this function only suitable for task ref?

Reply:
It works on any ref, but we seem to use it just for tasks. Also, the prefixes `R` (a "real" ref for something in the DB) and `D` (a dummy ref; nothing in the DB) are only relevant for tasks.